3_train/src/models.py

```diff
@@ -0,0 +1,351 @@
+
+"""
+This file is part of the accompanying code to our manuscript:
+
+Kratzert, F., Klotz, D., Shalev, G., Klambauer, G., Hochreiter, S., Nearing, G., "Benchmarking
+a Catchment-Aware Long Short-Term Memory Network (LSTM) for Large-Scale Hydrological Modeling".
+submitted to Hydrol. Earth Syst. Sci. Discussions (2019)
+
+You should have received a copy of the Apache-2.0 license along with the code. If not,
+see <https://opensource.org/licenses/Apache-2.0>
+"""
+
+class EALSTM(nn.Module):
+    """Implementation of the Entity-Aware-LSTM (EA-LSTM)
+
+    TODO: Include paper ref and latex equations
+
+    Parameters
+    ----------
+    input_size_dyn : int
+        Number of dynamic features, which are those, passed to the LSTM at each time step.
+    input_size_stat : int
+        Number of static features, which are those that are used to modulate the input gate.
+    hidden_size : int
+        Number of hidden/memory cells.
+    batch_first : bool, optional
+        If True, expects the batch inputs to be of shape [batch, seq, features] otherwise, the
+        shape has to be [seq, batch, features], by default True.
+    initial_forget_bias : int, optional
+        Value of the initial forget gate bias, by default 0
+
+    """
+
+    def __init__(self,
+                 input_size_dyn: int,
+                 input_size_stat: int,
+                 hidden_size: int,
+                 batch_first: bool = True,
+                 initial_forget_bias: int = 0):
+        super(EALSTM, self).__init__()
+
+        self.input_size_dyn = input_size_dyn
+        self.input_size_stat = input_size_stat
+        self.hidden_size = hidden_size
+        self.batch_first = batch_first
+        self.initial_forget_bias = initial_forget_bias
+
+        # create tensors of learnable parameters
+        self.weight_ih = nn.Parameter(torch.FloatTensor(input_size_dyn, 3 * hidden_size))
+        self.weight_hh = nn.Parameter(torch.FloatTensor(hidden_size, 3 * hidden_size))
+        self.weight_sh = nn.Parameter(torch.FloatTensor(input_size_stat, hidden_size))
+        self.bias = nn.Parameter(torch.FloatTensor(3 * hidden_size))
+        self.bias_s = nn.Parameter(torch.FloatTensor(hidden_size))
+
+        # initialize parameters
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        """Initialize all learnable parameters of the LSTM"""
+        nn.init.orthogonal_(self.weight_ih.data)
+        nn.init.orthogonal_(self.weight_sh)
+
+        weight_hh_data = torch.eye(self.hidden_size)
+        weight_hh_data = weight_hh_data.repeat(1, 3)
+        self.weight_hh.data = weight_hh_data
+
+        nn.init.constant_(self.bias.data, val=0)
+        nn.init.constant_(self.bias_s.data, val=0)
+
+        if self.initial_forget_bias != 0:
+            self.bias.data[:self.hidden_size] = self.initial_forget_bias
+
+    def forward(self, x_d: torch.Tensor, x_s: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
```

Comment:
Flow and operations check out 👍🏼 

The math is a strange idea at first, but I like it and think it makes sense to use a fixed input gate to modulate the candidate for short term memory. I was originally assuming the static features would be used only in a fully connected layer on top of the LSTM to sort of nudge the dynamic outputs, but I like this approach of them being more tightly coupled with the temporal dynamics.

Reply:
Yes, it's an intuitively pleasing concept after thinking it through!